lib/foreman/stored_result.rb

```diff
@@ -0,0 +1,43 @@
+module Foreman
+  class StoredResult
+    class TmpFileProvider
+      DIRECTORY = Rails.root.join('tmp')
+
+      def write(result_key, result)
```

Comment:
IMHO this is wrong, we should not support writing to files as this can break deployment that scale across multiple hosts, did you consider using the rails cache instead?

Reply:
I concur. Writing to files is introducing technical debt. What's wrong with streaming the data? Why do we have to do it async?

Reply:
This is another point to be hopefully opinionated :)
Rails.cache is not really persistent across instances eighter (you can change the implementation from database to file), files are than the same.
I have introduced this class, basically just to make an API to the StoredResult.
Now we can have these tree providers and I would like to end up at one:
1. Rails.cache
2. TempFile
3. Database - see `stored_results`

Reply:
@timogoebel there is big concern about the reports taking too much of a time to generate.
In my opinion "too much" would be one minute and more ( you are not really happy to wait for it, so you will let that to be sent to your inbox, what is going to be next PR of mine ).

Reply:
I just saw a talk at a ruby conference where somebody actually streamed the data directly from the database to the client. (csv genereated in sql, stream from postgresql via actioncontroller::metal to the client).

Reply:
> This is another point to be hopefully opinionated :)
What do you mean?

> Rails.cache is not really persistent across instances eighter (you can change the implementation from database to file), files are than the same.
Users with multi host behind a LB already knows that, and use a plugin such as foreman memcache to solve this issue.

> I have introduced this class, basically just to make an API to the StoredResult.
> Now we can have these tree providers and I would like to end up at one:
> 
> 1. Rails.cache
> 2. TempFile
> 3. Database - see `stored_results`

I would argue we can do 1 or 3, technically, 2 will work (we already do it in the smart proxy to store scap reports), but I think for foreman ability to scale we should not introduce it (e.g. it would require shared storage etc)
1. already offer abstraction between memory/file/shared service, and if your data is temporary, imho its a good solution, otherwise, consider using a metal controller to store the data as blob in your db.

Reply:
S3 compatible storage would be perfect for this, actually.

Reply:
Agreed :) or even if we just be running on kubernetes... :)

Reply:
@timogoebel well, it is not a problem to stream from ruby,  but to receive on the client without taking unecessary time from web worker. Even though it would be actually nice to think about the cases, when the result is not fitting to the memory (but it is a real corner case because there is a problem to process such reports on client anyway). And the most spent time is expected for rendering the report, the streaming should be relatively fast compared to the rendering.

Reply:
@ohadlevy well it would makes sense actually, but the problem is, that you are choosing the storage for cache for serving like a cache the best and you are not changing it for reports, so it can make sense for cache to be in the file, even though you got more instances, because you don't care that the content is processed on every worker, but this would make the polling to fail, if you are hitting the other instance, than it was processed on.

S3 compatible it is very simple to do actully, but not really good like a default option 🤣
(well you can make some reports, but buy and configure S3 first for them...)

Reply:
For me the best option would be to give a user choice between streaming through websockets or send by email. In that case you don't actually need any storage.

But we don't have websockets 😢

Reply:
Would you send a link to the file via mail or attach it to the mail?

Reply:
@timogoebel if it would be meant to avoid storing the results, than sending right away, so I didn't thought about sending link, but if we are storing the results already, we can give that choice actually, but the stored results are not meant to be available for more than 24 hours anyway :thinking:

Reply:
Ok, guys, I have chosen the database, with addition, that in the future, it would be nice to have an pulp storage option, if there happens to be one alongside the foreman.